cirq/circuits/circuit_operation.py

```diff
@@ -188,53 +262,97 @@ def _json_dict_(self):
             'qubit_map': sorted(self.qubit_map.items()),
             'measurement_key_map': self.measurement_key_map,
             'param_resolver': self.param_resolver,
+            'repetition_ids': self.repetition_ids,
         }
 
     @classmethod
     def _from_json_dict_(
-        cls, circuit, repetitions, qubit_map, measurement_key_map, param_resolver, **kwargs
+        cls,
+        circuit,
+        repetitions,
+        qubit_map,
+        measurement_key_map,
+        param_resolver,
+        repetition_ids,
+        **kwargs,
     ):
         return (
             cls(circuit)
             .with_qubit_mapping(dict(qubit_map))
             .with_measurement_key_mapping(measurement_key_map)
             .with_params(param_resolver)
-            .repeat(repetitions)
+            .repeat(repetitions, repetition_ids)
         )
 
     # Methods for constructing a similar object with one field modified.
 
     def repeat(
-        self,
-        repetitions: INT_TYPE,
+        self, repetitions: INT_TYPE, repetition_ids: Optional[List[str]] = None
     ) -> 'CircuitOperation':
         """Returns a copy of this operation repeated 'repetitions' times.
+         Each repetition instance will be identified by a single repetition_id.
 
         Args:
             repetitions: Number of times this operation should repeat. This
                 is multiplied with any pre-existing repetitions.
+            repetition_ids: List of IDs, one for each repetition. The length
+                should be equal to either `repetitions` or
+                `repetitions*self.repetitions`. In the latter case, these
```

Comment:
What is the use case for the `repetitions * self.repetitions` / override-all-IDs behavior? It seems like the `repetitions` / prefixing case should be sufficient.

Separately, we shouldn't rely on the size of this argument to determine its purpose. If we need both behaviors, we should have a separate argument (or method) for each.

Reply:
Since we assign default `repetition_ids` even when the circuit is non-measurement, without override-all-IDs the following will be true
```
circuit_op.repeat(2).repeat(3) != circuit_op.repeat(6)
circuit_op.repeat(2).repetition_ids == ['0-0', '1-0']
circuit_op.repeat(2, ['a', 'b']).repetition_ids == ['a-0', 'b-0']
```

The above can be handled by special casing based on if the base or input repetition_ids are "defaults"/None. But if we are overriding in some cases, might as well make that behavior available to the caller. For e.g. if `circuit_op.repeat(2).repeat(3) == circuit_op.repeat(6)`, it would be nice to be able to specify all 6 `repetition_ids` in the left-side construction.

But I agree that size-based might not be the best way to go about this. How about I add an `override_all_ids: Optional[bool]` arg? If specified, the `repetition_ids` size will be validated based on that. If unspecified, we choose the default value of `override_all_ids` based on `repetition_ids` and `self.repetition_ids`.

Reply:
Made the above modifications. Let me know if this looks better.

Reply:
In my mind, `op.repeat(2).repeat(3) != op.repeat(6)` is the correct behavior, since they represent different things. Conceptually, it's similar to how `[[0, 1], [2, 3]] != [0, 1, 2, 3]`.

Regarding `circuit_op.repeat(2).repetition_ids`: if the original `circuit_op` has an explicitly-defined ID of `0` for its single repetition, I think `['0-0', '1-0']` is correct. However, if no ID has been defined, we want `['0', '1']` instead. This behavior is similar to a raw value vs. a list with one element:
```
a = 0
a2 = [a, a]  # [0, 0]
a2[0] == a2[1] == 0

b = [0]
b2 = [b, b]  # [[0], [0]]
b2[0][0] == b2[1][0] == 0
```
With this in mind, our only edge case is the single-repetition, no-ID CircuitOperation; everything else follows the "prefix" case. By extension, we can drop `override_all_ids` and the extra complexity associated with it.

Reply:
Sure, I was more going for the `ControlledGate` type behavior we have. Chained controls with different number of qubits are automatically "flattened" and the final `ControlledGate` is equivalent to if all of them were applied at once.
 
But I guess `pow` semantics (repeat) don't need to be similar to that `addition` type operation.

Made the above change and simplified a couple of other things.